docs/topics/gateways.md

```diff
@@ -0,0 +1,322 @@
+Brigade ships with two gateways that translate webhook requests into Brigade
+events: The [GitHub gateway](github.md) and the
+[container registry gateway](dockerhub.md).
+These both provide HTTP-based listeners that receive incomming requests
+and generate Brigade events as a result.
+
+But the gateway system works with more than just webhooks.
+
+For example, the `brig` client also acts as a gateway. When you execute a `brig run`
+command, `brig` creates a Brigade event. By default, it emits an `exec` event. And
+Brigade itself processes this event no differently than it processes the GitHub
+or container registry events.
+
+There are no rules about what can be used as a trigger for an event. One could
+write a gateway that listens on a message queue, or runs as a chat bot, or watches
+files on a filesystem... any of these could be used to trigger a new Brigade event.
+
+The remainder of this guide explains how gateways work and how you can create custom
+gateways.
+
+## An Event Is A Secret
+
+The most important thing to understand about a Brigade event is that it is simply
+a [Kuberetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/)
+with special labels and data.
+
+When a new and appropriately labeled secret is creates in Kubernetes, the Brigade
+controller will read that secret and start a new Brigade worker to handle the event.
+Secrets have several characteristics that make them a great fit for this role:
+
+- They are designed to protect data (and we expect them to mature in this capacity)
+- They can be mounted as volumes and environment variables.
+- The payload of a secret is flexible
+- Secrets have been a stable part of the Kubernetes ecosystem since Kubernetes 1.2
+
+Because of these features, the Brigade system uses features as containers for
+bearing event information.
+
+### The Anatomy of a Brigade Event Secret
+
+Here is the structure of a Brigade event secret. It is annotated to explain what
+data belongs in what fields.
+
+```yaml
+# The main structure is a normal Kubernetes secret
+apiVersion: v1
+kind: Secret
+metadata:
+  # Every event has an automatically generated name. The main requirement of
+  # this is that it MUST BE UNIQUE.
+  name: example
+  # Brigade uses several labels to determine whether a secret carries a
+  # Brigade event.
+  labels:
+    # 'heritage: brigade' is mandatory, and signals that this is a Brigade event.
+    heritage: brigade
+
+    # This should point to the Brigade project ID in which this event is to be
+    # executed
+    project: brigade-1234567890
+
+    # This MUST be a unique ID. Where possible, it SHOULD be a ULID
+    # Substituting a UUID is fine, though some sorting functions won't be as expected
+    build: 01C1R2SYTYAR2WQ2DKNTW8SH08
+
+    # Commit is the commitish/reference for any associated VCS repository. By
+    # default, this should be `master` for Git.
+    commit: master
+
+    # 'component: build' is REQUIRED and tells brigade to start a new build.
```

Comment:
What if I don't want to build stuff? Maybe I just want to schedule tests on an already deployed release.

Reply:
We're using the term "build" to reference a "job", not necessarily to "build things".